// @shelby-protocol/solana-kit/dist/react/index.mjs
// src/common/constants.ts
import { defaultSolanaAuthenticationFunction } from "@aptos-labs/derived-wallet-solana";
import { Network as AptosNetwork } from "@aptos-labs/ts-sdk";
var Network = {
  SHELBYNET: AptosNetwork.SHELBYNET
};

// src/common/deriveAddress.ts
import {
  defaultSolanaAuthenticationFunction as defaultSolanaAuthenticationFunction2,
  SolanaDerivedPublicKey
} from "@aptos-labs/derived-wallet-solana";

// src/react/useStorageAccount.tsx
import {
  defaultSolanaAuthenticationFunction as defaultSolanaAuthenticationFunction3,
  SolanaDerivedPublicKey as SolanaDerivedPublicKey2,
  signAptosTransactionWithSolana
} from "@aptos-labs/derived-wallet-solana";
import { UserResponseStatus } from "@aptos-labs/wallet-standard";
import { PublicKey } from "@solana/web3.js";
import { useMemo } from "react";
function useStorageAccount(params) {
  const { client, wallet } = params;
  const domain = typeof window !== "undefined" ? window.location.host : void 0;
  const publicKey = wallet?.account?.address?.toString() ?? null;
  const derivedPublicKey = useMemo(() => {
    if (!publicKey || !domain) return null;
    try {
      return new SolanaDerivedPublicKey2({
        domain,
        solanaPublicKey: new PublicKey(publicKey),
        authenticationFunction: defaultSolanaAuthenticationFunction3
      });
    } catch {
      return null;
    }
  }, [publicKey, domain]);
  const storageAccountAddress = useMemo(
    () => derivedPublicKey?.authKey().derivedAddress() ?? null,
    [derivedPublicKey]
  );
  const signTransaction = async (params2) => {
    if (!publicKey) {
      throw new Error("Wallet not connected");
    }
    if (!derivedPublicKey) {
      throw new Error("Invalid Solana public key");
    }
    if (!storageAccountAddress) {
      throw new Error("Storage account address not found");
    }
    const rawTransaction = await client.aptos.transaction.build.simple({
      sender: storageAccountAddress,
      data: params2.data
    });
    if (!domain) {
      throw new Error("Domain is required in the browser");
    }
    const response = await signAptosTransactionWithSolana({
      solanaWallet: {
        publicKey: new PublicKey(publicKey),
        signMessage: wallet?.signMessage,
        signIn: wallet?.signIn,
        name: "Solana Wallet"
      },
      authenticationFunction: defaultSolanaAuthenticationFunction3,
      rawTransaction,
      domain
    });
    if (response.status !== UserResponseStatus.APPROVED || !("args" in response)) {
      throw new Error("User rejected the request.");
    }
    return {
      authenticator: response.args,
      rawTransaction
    };
  };
  const submitTransaction = async (params2) => {
    const { transaction, senderAuthenticator, transactionSubmitter } = params2;
    const submitted = await client.aptos.transaction.submit.simple({
      transaction,
      senderAuthenticator,
      transactionSubmitter: transactionSubmitter ?? null
    });
    const committed = await client.aptos.waitForTransaction({
      transactionHash: submitted.hash
    });
    return { hash: committed.hash };
  };
  const signAndSubmitTransaction = async (params2) => {
    const { data, transactionSubmitter } = params2;
    const { authenticator, rawTransaction } = await signTransaction({ data });
    return submitTransaction({
      transaction: rawTransaction,
      senderAuthenticator: authenticator,
      transactionSubmitter
    });
  };
  return {
    storageAccountAddress,
    signTransaction,
    submitTransaction,
    signAndSubmitTransaction
  };
}
export {
  Network,
  useStorageAccount
};
