// @shelby-protocol/solana-kit/dist/node/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _createStarExport(obj) { Object.keys(obj) .filter((key) => key !== "default" && key !== "__esModule") .forEach((key) => { if (exports.hasOwnProperty(key)) { return; } Object.defineProperty(exports, key, {enumerable: true, configurable: true, get: () => obj[key]}); }); }// src/node/index.ts
var _node = require('@shelby-protocol/sdk/node'); _createStarExport(_node);

// src/common/constants.ts
var _derivedwalletsolana = require('@aptos-labs/derived-wallet-solana');
var _tssdk = require('@aptos-labs/ts-sdk');
var Network = {
  SHELBYNET: _tssdk.Network.SHELBYNET
};

// src/common/deriveAddress.ts





// src/node/shelby.ts
var _gasstationclient = require('@aptos-labs/gas-station-client');




// src/node/storageAccount.ts

var ShelbyStorageAccount = class extends _derivedwalletsolana.SolanaDerivedAccount {
};

// src/node/shelby.ts
var Shelby = class extends _node.ShelbyClient {
  constructor(params) {
    const shelbyClientConfig = {
      network: params.network,
      apiKey: params.apiKey
    };
    if (params.gasStationApiKey) {
      shelbyClientConfig.aptos = {
        pluginSettings: {
          TRANSACTION_SUBMITTER: new (0, _gasstationclient.GasStationTransactionSubmitter)({
            network: params.network,
            apiKey: params.gasStationApiKey
          })
        }
      };
    }
    super(shelbyClientConfig);
    this.connection = params.connection;
  }
  /**
   * Create a Shelby storage account
   *
   * @param solanaKeypair - The Solana keypair
   * @param domain - The domain of the storage account
   * @returns The Shelby storage account
   *
   * @example
   * ```typescript
   * const storageAccount = shelbyClient.createStorageAccount(solanaKeypair, domain);
   * ```
   */
  createStorageAccount(solanaKeypair, domain) {
    return new ShelbyStorageAccount({ solanaKeypair, domain });
  }
};



exports.Network = Network; exports.Shelby = Shelby;
