// @shelby-protocol/solana-kit/dist/node/index.mjs
// src/node/index.ts
export * from "@shelby-protocol/sdk/node";

// src/common/constants.ts
import { defaultSolanaAuthenticationFunction } from "@aptos-labs/derived-wallet-solana";
import { Network as AptosNetwork } from "@aptos-labs/ts-sdk";
var Network = {
  SHELBYNET: AptosNetwork.SHELBYNET
};

// src/common/deriveAddress.ts
import {
  defaultSolanaAuthenticationFunction as defaultSolanaAuthenticationFunction2,
  SolanaDerivedPublicKey
} from "@aptos-labs/derived-wallet-solana";

// src/node/shelby.ts
import { GasStationTransactionSubmitter } from "@aptos-labs/gas-station-client";
import {
  ShelbyClient
} from "@shelby-protocol/sdk/node";

// src/node/storageAccount.ts
import { SolanaDerivedAccount } from "@aptos-labs/derived-wallet-solana";
var ShelbyStorageAccount = class extends SolanaDerivedAccount {
};

// src/node/shelby.ts
var Shelby = class extends ShelbyClient {
  constructor(params) {
    const shelbyClientConfig = {
      network: params.network,
      apiKey: params.apiKey
    };
    if (params.gasStationApiKey) {
      shelbyClientConfig.aptos = {
        pluginSettings: {
          TRANSACTION_SUBMITTER: new GasStationTransactionSubmitter({
            network: params.network,
            apiKey: params.gasStationApiKey
          })
        }
      };
    }
    super(shelbyClientConfig);
    this.connection = params.connection;
  }
  /**
   * Create a Shelby storage account
   *
   * @param solanaKeypair - The Solana keypair
   * @param domain - The domain of the storage account
   * @returns The Shelby storage account
   *
   * @example
   * ```typescript
   * const storageAccount = shelbyClient.createStorageAccount(solanaKeypair, domain);
   * ```
   */
  createStorageAccount(solanaKeypair, domain) {
    return new ShelbyStorageAccount({ solanaKeypair, domain });
  }
};
export {
  Network,
  Shelby
};
