// @aptos-labs/gas-station-client/src/GasStationClient.ts
import { AccountAuthenticator, AnyRawTransaction, Network } from "@aptos-labs/ts-sdk";
import { type Config, createClient, createConfig } from "./generated/client";
import { signAndSubmitTransaction } from "./generated";

export interface GasStationClientOptions extends Config {
  network?: Network;
  env?: "prod" | "staging";
  apiKey?: string;
}

export interface SignAndSubmitTransactionBody {
  transaction: AnyRawTransaction;
  senderAuthenticator: AccountAuthenticator;
  secondaryAuthenticators?: AccountAuthenticator[];
  recaptchaToken?: string;
}

export class GasStationClient {
  readonly client: ReturnType<typeof createClient>;

  constructor(options: GasStationClientOptions) {
    const { apiKey, network = Network.MAINNET, env = "prod" } = options;
    let baseUrl = options.baseUrl;
    if (!baseUrl) {
      const envSegment = env !== "prod" ? `.${env}` : "";
      baseUrl = `https://api.${network}${envSegment}.aptoslabs.com/gs/v1`;
    }
    const headers = new Headers(options.headers as HeadersInit | undefined);
    if (apiKey) {
      headers.set("Authorization", `Bearer ${apiKey}`);
    }
    const config = createConfig({
      ...options,
      baseUrl,
      headers,
    });

    this.client = createClient(config);
  }

  async signAndSubmitTransaction(body: SignAndSubmitTransactionBody) {
    const { transaction, senderAuthenticator, secondaryAuthenticators, recaptchaToken } = body;

    const transactionBytes = Array.from(transaction.bcsToBytes());
    const senderAuth = Array.from(senderAuthenticator.bcsToBytes());
    const additionalSignersAuth = secondaryAuthenticators?.map((authenticator) =>
      Array.from(authenticator.bcsToBytes()),
    );

    const result = await signAndSubmitTransaction({
      client: this.client,
      throwOnError: true,
      body: {
        transactionBytes,
        senderAuth,
        additionalSignersAuth,
        recaptchaToken,
      },
    });

    return result.data;
  }
}
