// @aptos-labs/gas-station-client/src/GasStationTransactionSubmitter.ts
import {
  AptosConfig,
  InputSubmitTransactionData,
  PendingTransactionResponse,
  TransactionResponseType,
  TransactionSubmitter,
} from "@aptos-labs/ts-sdk";
import { GasStationClient, GasStationClientOptions } from "./GasStationClient";

export class GasStationTransactionSubmitter implements TransactionSubmitter {
  readonly client: GasStationClient;

  constructor(clientOrOptions: GasStationClient | GasStationClientOptions) {
    this.client =
      clientOrOptions instanceof GasStationClient
        ? clientOrOptions
        : new GasStationClient(clientOrOptions);
  }

  /**
   * This is an implementation of TransactionSubmitter interface that lets you use
   * this with the TS SDK plugin system. See the package README for more details.
   *
   * @param args - The arguments for submitting the transaction.
   * @param args.aptosConfig - The configuration for connecting to the Aptos network
   * (unused in gas station).
   * @param args.transaction - The Aptos transaction data to be submitted.
   * @param args.senderAuthenticator - The account authenticator of the transaction
   * sender.
   * @param args.secondarySignerAuthenticators - Optional. Authenticators for
   * additional signers (ignored for now).
   *
   * To provide the recaptcha token, you can pass it in the pluginParams object.
   * ```
   * const response = await client.submitTransaction({
   *   transaction,
   *   senderAuthenticator,
   *   // This part is only required if you have recaptcha enabled.
   *   pluginParams: { recaptchaToken },
   * });
   *
   * @returns PendingTransactionResponse - The response containing the status of the
   * submitted transaction. NOTE: We do not populate `payload` for now, it will be an
   * empty object.
   */
  async submitTransaction(
    args: {
      aptosConfig: AptosConfig;
    } & Omit<InputSubmitTransactionData, "transactionSubmitter">,
  ): Promise<PendingTransactionResponse> {
    const { transaction, senderAuthenticator, additionalSignersAuthenticators, pluginParams } =
      args;

    const recaptchaToken =
      typeof pluginParams?.recaptchaToken === "string" ? pluginParams.recaptchaToken : undefined;

    if (transaction.feePayerAddress === undefined) {
      throw new Error(
        "Transaction must have a fee payer for gas station plugin to work, set withFeePayer: true",
      );
    }

    const { transactionHash } = await this.client.signAndSubmitTransaction({
      transaction,
      senderAuthenticator,
      secondaryAuthenticators: additionalSignersAuthenticators,
      recaptchaToken,
    });

    // Build PendingTransactionResponse from SimpleTransaction.rawTransaction and the
    // gas station response.
    const rawTxn = transaction.rawTransaction;

    return {
      type: TransactionResponseType.Pending,
      hash: transactionHash,
      sender: rawTxn.sender.toString(),
      sequence_number: rawTxn.sequence_number.toString(),
      max_gas_amount: rawTxn.max_gas_amount.toString(),
      gas_unit_price: rawTxn.gas_unit_price.toString(),
      expiration_timestamp_secs: rawTxn.expiration_timestamp_secs.toString(),
      // We don't have a very safe way to access the payload, just return nothing for now.
      payload: {} as any,
    };
  }
}
